File: leetcode/completed/110.cpp
#include <math.h>
#include <algorithm>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int dfs(TreeNode* node) {
            if (node == nullptr) return 0;

            int left = dfs(node->left);
            int right = dfs(node->right);

            if (left == -1 || right == -1 || abs(left - right) > 1) return -1;

            return max(left, right)+1;
        }     

    bool isBalanced(TreeNode* root) {
        return dfs(root) != -1;
    }
};

File: leetcode/completed/110.py
from typing import Optional

# If abs(depth_left - depth_right) > 1, then return -1.
# Else, return maximum depth between left and right.
# Using -1 as signal, if dfs returns -1, it means res = False, so return -1 all the way.

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root: return True

        def dfs(node):
            if not node: return 0

            left = dfs(node.left)
            right = dfs(node.right)

            if left == -1 or right == -1 or abs(left-right) > 1:
                return -1
            
            return max(left, right) + 1
        
        return dfs(root) != -1
        

            